### PR TITLE
 Preciso Bid adapter : Initial release

### PR DESCRIPTION
**Pull request by PrecisoSRL**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

### CODE REVIEW

**Comment by PrecisoSRL** [comment]
doc :  https://github.com/prebid/prebid.github.io/pull/4639

File: modules/precisoBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+    if (validBidRequests[0].userIdAsEids) {
+      const eids = { eids: validBidRequests[0].userIdAsEids };
+      if (request.user && request.user.ext) {
+        request.user.ext = { ...request.user.ext, ...eids };
+      } else {
+        request.user = {ext: eids};
+      }
+    }
+
+    const ortb2Params = bidderRequest?.ortb2 || {};
```
Comment (ncolletti) [inline modules/precisoBidAdapter.js:87]: Could you update to use `getConfig('ortb2')` to access this data?

File: modules/precisoBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+ * @returns {object} Prebid bidObject.native
+ */
+function interpretNativeAd(adm) {
+  const native = JSON.parse(adm).native;
+  const result = {
+    clickUrl: encodeURI(native.link.url),
+    impressionTrackers: native.imptrackers
+  };
+  native.assets.forEach(asset => {
+    switch (asset.id) {
```
Comment (ncolletti) [inline modules/precisoBidAdapter.js:483]: recommend using a default statement here

File: modules/precisoBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+  buildRequests: function (validBidRequests, bidderRequest) {
+    // convert Native ORTB definition to old-style prebid native definition
+    validBidRequests = convertOrtbRequestToProprietaryNative(validBidRequests);
+
+    const request = {
+      id: bidderRequest.bidderRequestId,
+      imp: validBidRequests.map(slot => mapImpression(slot, bidderRequest)),
+      site: mapSite(validBidRequests, bidderRequest),
+      cur: DEFAULT_CURRENCY_ARR,
+      test: validBidRequests[0].params.test || 0,
```
Comment (ncolletti) [inline modules/precisoBidAdapter.js:58]: Can you update the adapter to use `deepAccess` in places where you are referencing a nested property?

File: modules/precisoBidAdapter.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	           banner: {
+                       sizes: [[300, 250]],
+                   }
+               },
+               bids: [
+                   {
+                       bidder: "preciso",
+                       params: {
+                           region: 'prebid-eu',
+                           publisherId: 'PREBID_TEST_ID',
```
Comment (ncolletti) [inline modules/precisoBidAdapter.md:28]: Please update the publisherId to a working id for testing, thank you!

File: modules/precisoBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+import {getOrigin} from '../libraries/getOrigin/index.js';
+import {BANNER, NATIVE, VIDEO} from '../src/mediaTypes.js';
+import {registerBidder} from '../src/adapters/bidderFactory.js';
+import {includes} from '../src/polyfill.js';
+import {convertOrtbRequestToProprietaryNative} from '../src/native.js';
+import {config} from '../src/config.js';
+
+const BIDDER_CODE = 'preciso';
+const REGIONS = ['prebid-eu', 'prebid-us', 'prebid-asia'];
+const ENDPOINT_URL = 'mndtrk.com/bidder/prebid/bids';
```
Comment (ncolletti) [inline modules/precisoBidAdapter.js:11]: All regions are returning an ERR_NAME_NOT_RESOLVED error on this endpoint URL. Please resolve so we can continue testing your adapter. Thank you!

File: modules/precisoBidAdapter.js (hunk 1 above)
Comment (PrecisoSRL) [reply modules/precisoBidAdapter.js:87]: Taking the rtbhouseBidAdapter.js as a reference, the precisoBidAdapter.js was created. We noticed that getConfig('ortb2') is not used in the rtbhouseBidAdapter.js file. Could you please explain how to use getConf('ortb2)?

File: modules/precisoBidAdapter.md (hunk 4 above)
Comment (PrecisoSRL) [reply modules/precisoBidAdapter.md:28]: We referred to the RTB House bid adapter while creating ours. The RTB House bidder used The same publisher ID. Is the publisher ID incorrect for testing?

File: modules/precisoBidAdapter.js (hunk 1 above)
Comment (ncolletti) [reply modules/precisoBidAdapter.js:87]: More information can be found at the documentation, it retrieves ortb2 data from the auction: https://docs.prebid.org/dev-docs/publisher-api-reference/setConfig.html#setConfig-fpd

File: modules/precisoBidAdapter.md (hunk 4 above)
Comment (ncolletti) [reply modules/precisoBidAdapter.md:28]: Ideally we'd like to see a publisher Id that your endpoint will return a bid with so that we can confirm your adapter packages requests and responds to your bids properly.

File: modules/precisoBidAdapter.js (hunk 1 above)
Comment (ncolletti) [reply modules/precisoBidAdapter.js:87]: @PrecisoSRL We generally ask new adapters to use the getConfig method for pulling ortb2 data, not to directly access it from the bidrequest. Can you update, this should be no downstream change to the data you receive.

File: modules/precisoBidAdapter.md (hunk 4 above)
Comment (ncolletti) [reply modules/precisoBidAdapter.md:28]: This still needs to be updated, thank you

**Comment by ncolletti** [comment]
We would expect to see your endpoint return bids when testing using the local environment within the prebid.js repo. After the recent updates and using the parameters in your markdown file, I am currently I'm seeing CORS errors on your AD_URL endpoint

**Comment by ncolletti** [comment]
Checked back in, I am still seeing a CORS error from your endpoint when testing with the supplied params and endpoint from this PR. Could you please look into updating your server so it returns a valid bid to complete my testing? Thank you

**Comment by PrecisoSRL** [comment]
Thank you for updating me.  Currently, we are working on fixing the issue. As soon as it is completed, we will update here.

**Comment by PrecisoSRL** [comment]
On the particular server, we deployed our bidder application. Would you mind testing again?

**Comment by ncolletti** [comment]
@PrecisoSRL Looks like there is still a CORS error. To test on your end, you can run a `gulp serve` and then navigate to `http://localhost:9999/integrationExamples/gpt/hello_world.html` and enter your bidder params from your markdown file.

**Comment by PrecisoSRL** [comment]
thank you for updating me. we will check and fix the issue  as soon as possible

**Comment by ncolletti** [comment]
@PrecisoSRL Seeing a 204 now, which is good progress. Now can you make sure your example params in your markdown file return test responses for these specific params? The expectation is that a User can try out your adapter and see it working in the their Prebid auction. We'd like to see this before we can move forward with approving the PR. Thank you!

File: integrationExamples/gpt/hello_world.html [hunk 6]
```diff
@@ -24,9 +24,12 @@
             },
             // Replace this object to test a new Adapter!
             bids: [{
-                bidder: 'appnexus',
+                bidder: 'preciso',
```
Comment (patmmccann) [inline integrationExamples/gpt/hello_world.html:27]: you don't get to change the hello world to your own, you can add a new integration example

**Review by patmmccann — DISMISSED** [review]

File: integrationExamples/gpt/hello_world.html [hunk 7]
```diff
@@ -26,7 +26,9 @@
             bids: [{
                 bidder: 'appnexus',
                 params: {
-                    placementId: 13144370
+                    host: 'prebid',
+                    region: 'prebid-eu',
+                    traffic: 'banner'
```
Comment (ncolletti) [inline integrationExamples/gpt/hello_world.html:31]: revert this change

Comment (PrecisoSRL) [reply integrationExamples/gpt/hello_world.html]: I have reverted the changes. I'm sorry for not having changed that.

**Review by ncolletti — APPROVED** [review]
looks good!
